app/forms/new_deputy_assignment_form.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+class NewDeputyAssignmentForm
+  class IdentityServiceError < StandardError; end
+  class IdentityNotFound < StandardError; end
+  class AlreadyAssigned < StandardError; end
+  class ErrorCreatingUser < StandardError; end
+
+  include ActiveModel::Model
+
+  attr_accessor :primary,
+                :deputy_webaccess_id
+
+  validates :primary,
+            presence: true
+
+  validates :deputy_webaccess_id,
+            presence: true,
+            format: {
+              with: /\A[a-z0-9]+\z/i,
+              allow_blank: true
+            }
+
+  def save
+    return false unless valid?
+
+    begin
+      deputy = find_or_initialize_deputy(webaccess_id: deputy_webaccess_id)
+      create_deputy_assignment(primary: primary, deputy: deputy)
+    rescue IdentityServiceError
+      errors.add(:base, :identity_service_error)
+    rescue IdentityNotFound
+      errors.add(:deputy_webaccess_id, :not_found)
+    rescue AlreadyAssigned
+      errors.add(:deputy_webaccess_id, :already_assigned)
+    rescue ErrorCreatingUser
+      errors.add(:base, :error_creating_user)
+    rescue StandardError
+      errors.add(:base, :unknown_error)
+    end
+
+    errors.empty?
+  end
+
+  private
+
+    def find_or_initialize_deputy(webaccess_id:)
+      User.find_or_initialize_by(webaccess_id: webaccess_id) do |new_user|
+        psu_identity = query_psu_identity(webaccess_id)
+        raise IdentityNotFound if psu_identity.blank?
+
+        new_user.first_name = psu_identity.given_name
+        new_user.last_name = psu_identity.family_name
+        new_user.psu_identity = psu_identity
+
+        raise ErrorCreatingUser unless new_user.valid?
```

Comment:
I could add errors here, but I still need to raise because I need a way to short-circuit the `#save` method if a User cannot be created, i.e. the calling method should not proceed to try to build a DeputyAssignment if the deputy User is invalid and can't be saved.

Now Avdi Grimm claims that you should [never use exceptions for control flow](https://www.rubytapas.com/2013/06/24/episode-110-catch-and-throw/) and that you should use `try / catch` for this purpose instead. The thing is, though, I've never _seen anyone actually ever do that._ 🤣